## Loading and looking at the dataset

In [1]:
import pandas as pd 

df = pd.read_csv("../data/Fraudulent_E-Commerce_Transaction_Data.csv")

#by default head() displays the first five rows of the dataswt
df.head()


,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour
0,c12e07a0-8a06-4c0d-b5cc-04f3af688570,8ca9f102-02a4-4207-ab63-484e83a1bdf0,42.32,2024-03-24 23:42:43,PayPal,electronics,1,40,East Jameshaven,desktop,110.87.246.85,5399 Rachel Stravenue Suite 718\nNorth Blakebu...,5399 Rachel Stravenue Suite 718\nNorth Blakebu...,0,282,23
1,7d187603-7961-4fce-9827-9698e2b6a201,4d158416-caae-4b09-bd5b-15235deb9129,301.34,2024-01-22 00:53:31,credit card,electronics,3,35,Kingstad,tablet,14.73.104.153,"5230 Stephanie Forge\nCollinsbury, PR 81853","5230 Stephanie Forge\nCollinsbury, PR 81853",0,223,0
2,f2c14f9d-92df-4aaf-8931-ceaf4e63ed72,ccae47b8-75c7-4f5a-aa9e-957deced2137,340.32,2024-01-22 08:06:03,debit card,toys & games,5,29,North Ryan,desktop,67.58.94.93,"195 Cole Oval\nPort Larry, IA 58422","4772 David Stravenue Apt. 447\nVelasquezside, ...",0,360,8
3,e9949bfa-194d-486b-84da-9565fca9e5ce,b04960c0-aeee-4907-b1cd-4819016adcef,95.77,2024-01-16 20:34:53,credit card,electronics,5,45,Kaylaville,mobile,202.122.126.216,"7609 Cynthia Square\nWest Brenda, NV 23016","7609 Cynthia Square\nWest Brenda, NV 23016",0,325,20
4,7362837c-7538-434e-8731-0df713f5f26d,de9d6351-b3a7-4bc7-9a55-8f013eb66928,77.45,2024-01-16 15:47:23,credit card,clothing,5,42,North Edwardborough,desktop,96.77.232.76,"2494 Robert Ramp Suite 313\nRobinsonport, AS 5...","2494 Robert Ramp Suite 313\nRobinsonport, AS 5...",0,116,15


In [2]:
#displays the number of rows and columns in the dataset in a tuple format 
df.shape

(23634, 16)

In [3]:
#displays the statistical summary of the dataset like mean , std , quartiles , count
df.describe()

,Transaction Amount,Quantity,Customer Age,Is Fraudulent,Account Age Days,Transaction Hour
count,23634.000000,23634.000000,23634.000000,23634.000000,23634.000000,23634.000000
mean,229.367099,3.000550,34.560210,0.051705,178.660531,11.266015
std,282.046669,1.419663,10.009471,0.221436,107.388682,6.980659
min,10.000000,1.000000,-2.000000,0.000000,1.000000,0.000000
25%,69.070000,2.000000,28.000000,0.000000,84.000000,5.000000
50%,151.415000,3.000000,35.000000,0.000000,178.000000,11.000000
75%,296.127500,4.000000,41.000000,0.000000,272.000000,17.000000
max,9716.500000,5.000000,73.000000,1.000000,365.000000,23.000000


In [4]:
#Displays a concise summary of the dataset 
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 23634 entries, 0 to 23633
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Transaction ID      23634 non-null  str    
 1   Customer ID         23634 non-null  str    
 2   Transaction Amount  23634 non-null  float64
 3   Transaction Date    23634 non-null  str    
 4   Payment Method      23634 non-null  str    
 5   Product Category    23634 non-null  str    
 6   Quantity            23634 non-null  int64  
 7   Customer Age        23634 non-null  int64  
 8   Customer Location   23634 non-null  str    
 9   Device Used         23634 non-null  str    
 10  IP Address          23634 non-null  str    
 11  Shipping Address    23634 non-null  str    
 12  Billing Address     23634 non-null  str    
 13  Is Fraudulent       23634 non-null  int64  
 14  Account Age Days    23634 non-null  int64  
 15  Transaction Hour    23634 non-null  int64  
dtypes: float64(1), 

## Cleaning the dataset

#### Looking for null values

In [5]:
df.isnull().sum()

Transaction ID        0
Customer ID           0
Transaction Amount    0
Transaction Date      0
Payment Method        0
Product Category      0
Quantity              0
Customer Age          0
Customer Location     0
Device Used           0
IP Address            0
Shipping Address      0
Billing Address       0
Is Fraudulent         0
Account Age Days      0
Transaction Hour      0
dtype: int64

#### Looking for duplicates

In [6]:
df.duplicated().sum()

np.int64(0)

##### Looking at the output above , there are no null and duplicate values in the dataset, so there is no need to drop (dropna()) or fill (fillna()) for null or use (drop_duplicates()) for duplicates

##### counting the number of values that are fraudulent/risky using the label given in the dataset

In [7]:
df["Is Fraudulent"].value_counts()

Is Fraudulent
0    22412
1     1222
Name: count, dtype: int64

##### 0 indicates normal orders and 1 indicates fraudulent orders

#### Calculating the percentage of risky orders


In [60]:
normal_order_percentage=(22412/23634)*100
risky_order_percentage=(1222/23634)*100
print(f"%age of normal orders:{round(normal_order_percentage,2)}")
print(f"%age of risky orders:{round(risky_order_percentage,2)}")

%age of normal orders:94.83
%age of risky orders:5.17


##### in the above dataset , there were no missing values or duplicate records. The target variable "Is Fraudulent" is imbalanced, with only 5.17% fraudulent transactions, which should be considered when selecting evaluation metrics for machine learning models.

#### Looking for invalid values in dataset

In [61]:
# A Transaction Amount can never be negative
df[df["Transaction Amount"]<0]

,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour


In [62]:
# A customer age can never be negative
df[df["Customer Age"]<0]

,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour
12919,7fab7db0-e4d2-4b51-98f3-bb17b5153224,e0345e5b-d91c-4851-b9ae-5170c9b8e7cc,190.76,2024-01-05 15:02:49,PayPal,health & beauty,3,-2,Cohenmouth,desktop,134.95.208.38,Unit 2353 Box 9474\nDPO AA 79121,Unit 2353 Box 9474\nDPO AA 79121,0,253,15


##### there is one invalid value in the 'Customer Age' column i.e. -2 and an age can never be negative . Since , it is only one record we will replace it with median age , to preserve the integrity of records and not lose any data point

In [63]:
median_age=df["Customer Age"].median()

# Replacing the invalid value with median age 
df.loc[df["Customer Age"]<0,"Customer Age"]=median_age

df[df["Customer Age"]<0]

,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour


In [64]:
# Quantity of a product can never be negative
df[df["Quantity"]<0]

,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour


In [65]:
#Looking for empty or invalid Transaction ID or Customer ID
df[df["Transaction ID"].isna()]

df[df["Customer ID"].isna()]

,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour


In [66]:
# Label should contain only binary value (0 and 1)

df["Is Fraudulent"].unique()

array([0, 1])

In [67]:
#Account Age can never be negative
df[df["Account Age Days"]<0]

,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour


In [68]:
# checking for invalid inputs in categorical columns

print(df["Payment Method"].value_counts())

print(df["Device Used"].value_counts())

print(df["Product Category"].value_counts())


Payment Method
debit card       5952
credit card      5923
PayPal           5899
bank transfer    5860
Name: count, dtype: int64
Device Used
desktop    7923
mobile     7881
tablet     7830
Name: count, dtype: int64
Product Category
home & garden      4786
electronics        4748
toys & games       4730
clothing           4699
health & beauty    4671
Name: count, dtype: int64


In [69]:
# Transaction Hour will always be between 0-23
df[(df['Transaction Hour'] < 0) | (df['Transaction Hour'] > 23)]

,Transaction ID,Customer ID,Transaction Amount,Transaction Date,Payment Method,Product Category,Quantity,Customer Age,Customer Location,Device Used,IP Address,Shipping Address,Billing Address,Is Fraudulent,Account Age Days,Transaction Hour


In [70]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 23634 entries, 0 to 23633
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Transaction ID      23634 non-null  str    
 1   Customer ID         23634 non-null  str    
 2   Transaction Amount  23634 non-null  float64
 3   Transaction Date    23634 non-null  str    
 4   Payment Method      23634 non-null  str    
 5   Product Category    23634 non-null  str    
 6   Quantity            23634 non-null  int64  
 7   Customer Age        23634 non-null  int64  
 8   Customer Location   23634 non-null  str    
 9   Device Used         23634 non-null  str    
 10  IP Address          23634 non-null  str    
 11  Shipping Address    23634 non-null  str    
 12  Billing Address     23634 non-null  str    
 13  Is Fraudulent       23634 non-null  int64  
 14  Account Age Days    23634 non-null  int64  
 15  Transaction Hour    23634 non-null  int64  
dtypes: float64(1), 

##### Data quality checks revealed no missing values and no duplicate records. One invalid value was identified in the Customer Age column (negative age), which was replaced with the median age. All other numerical fields were checked for impossible values, and no additional invalid records were found.